src/main/java/bio/terra/service/snapshotbuilder/CriteriaQueryUtils.java

```diff
@@ -0,0 +1,146 @@
+import bio.terra.service.snapshotbuilder.query.filtervariable.BinaryFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.BooleanAndOrFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.FunctionFilterVariable;
+import bio.terra.service.snapshotbuilder.query.filtervariable.SubQueryFilterVariable;
+import java.util.List;
+import org.apache.arrow.util.VisibleForTesting;
+
+public class CriteriaQueryUtils {
+  @VisibleForTesting
+  public static BooleanAndOrFilterVariable generateFilterForRangeCriteria(
+      SnapshotBuilderProgramDataRangeCriteria rangeCriteria) {
+    TablePointer tablePointer = TablePointer.fromTableName("person");
+    TableVariable tableVariable = TableVariable.forPrimary(tablePointer);
+    return new BooleanAndOrFilterVariable(
+        BooleanAndOrFilterVariable.LogicalOperator.AND,
+        List.of(
+            new BinaryFilterVariable(
+                new FieldVariable(
+                    new FieldPointer(tablePointer, rangeCriteria.getName()), tableVariable),
+                BinaryFilterVariable.BinaryOperator.GREATER_THAN_OR_EQUAL,
+                new Literal(rangeCriteria.getLow().doubleValue())),
+            new BinaryFilterVariable(
+                new FieldVariable(
+                    new FieldPointer(tablePointer, rangeCriteria.getName()), tableVariable),
+                BinaryFilterVariable.BinaryOperator.LESS_THAN_OR_EQUAL,
+                new Literal(rangeCriteria.getHigh().doubleValue()))));
+  }
+
+  @VisibleForTesting
+  public static FunctionFilterVariable generateFilterForListCriteria(
+      SnapshotBuilderProgramDataListCriteria listCriteria) {
+    TablePointer tablePointer = TablePointer.fromTableName("person");
+    TableVariable tableVariable = TableVariable.forPrimary(tablePointer);
+    TablePointer conceptPointer = TablePointer.fromTableName("concept");
+    TableVariable conceptPointerVariable =
+        TableVariable.forJoined(
+            conceptPointer,
+            "concept_id",
+            new FieldVariable(
+                new FieldPointer(tablePointer, listCriteria.getName()), tableVariable));
+    return new FunctionFilterVariable(
+        FunctionFilterVariable.FunctionTemplate.IN,
+        new FieldVariable(new FieldPointer(conceptPointer, "concept_id"), conceptPointerVariable),
+        listCriteria.getValues().stream()
+            .map(value -> new Literal(value.doubleValue()))
+            .toArray(Literal[]::new));
+  }
+
+  private record OccurrenceTable(String tableName, String idColumnName) {}
+
+  private static OccurrenceTable getOccurrenceTableFromDomainCriteria(
+      SnapshotBuilderDomainCriteria domainCriteria) {
+    // TODO: Load this information from snapshot builder settings, rather than hardcode
+    return switch (domainCriteria.getDomainName()) {
+      case "Condition" -> new OccurrenceTable("condition_occurrence", "condition_concept_id");
```

Comment:
If it's ok to re-use occurrence tables than using a static `Map` is probably a better way to do this. Something like
```
  private static final Map<String, OccurrenceTable> OCCURRENCE_TABLE_MAP =
      Map.of(
          "Condition", new OccurrenceTable("condition_occurrence", "condition_concept_id"),
          "Measurement", new OccurrenceTable("measurement", "measurement_concept_id"),
          "Visit", new OccurrenceTable("visit_occurrence", "visit_concept_id"),
          "Procedure", new OccurrenceTable("procedure_occurrence", "procedure_concept_id"),
          "Observation", new OccurrenceTable("observation", "observation_concept_id"),
          "Device", new OccurrenceTable("device_exposure", "device_concept_id"),
          "Drug", new OccurrenceTable("drug_exposure", "drug_concept_id"));

  private static OccurrenceTable getOccurrenceTableFromDomainCriteria(
      SnapshotBuilderDomainCriteria domainCriteria) {
    // TODO: Load this information from snapshot builder settings, rather than hardcode
    var occurrenceTable = OCCURRENCE_TABLE_MAP.get(domainCriteria.getDomainName());
    if (occurrenceTable == null) {
      throw new BadRequestException(
          String.format("Domain %s is not found in dataset", domainCriteria.getDomainName()));
    }
    return occurrenceTable;
  }
```